### PR TITLE
Add fee limit to `pay` command

### PR DESCRIPTION
**Pull request by ZmnSCPxj**
Fixes; #717

### CODE REVIEW

**Comment by cdecker** [comment]
My alarm bells start ringing when doubles are used anywhere near financial data, but I think in this case we can make an exception since the worst that can happen is overpaying the fee slightly, or being too restrictive. It's probably what users will feel comfortable anyway.

ACK dbad6ff

**Comment by ZmnSCPxj** [comment]
I understand.  However:

1.  We do not accumulate multiple additions or multiplications.  The result of multiplying and then dividing is compared to a user-provided limiting value rather than used in further computation.
2.  Given the 42mBTC limit, the number of bits used in the mantissa is 32.  Multiply by 100.0 adds about 7 bits, and 100.0 is representable exactly in a double as 1.1001 x 2^6 (mantissa in binary) .Granted there is a division afterwards, though.

I would have preferred the GMP ratio type, but that would have pulled in a rather large dependency for only this feature.

File: lightningd/pay.c [hunk 1]
```diff
@@ -763,7 +811,11 @@ static void json_pay(struct command *cmd,
 static const struct json_command pay_command = {
 	"pay",
 	json_pay,
-	"Send payment specified by {bolt11} with optional {msatoshi} (if and only if {bolt11} does not have amount), {description} (required if {bolt11} uses description hash) and {riskfactor} (default 1.0)"
+	"Send payment specified by {bolt11} with optional {msatoshi} "
+	"(if and only if {bolt11} does not have amount), "
+	"{description} (required if {bolt11} uses description hash), "
+	"{riskfactor} (default 1.0), and "
+	"{maxfee} (default 0.5) the maximum acceptable fee as a percentage (e.g. 0.5 => 0.5%)"
```
Comment (rustyrussell) [inline lightningd/pay.c]: I mildly prefer to call it `maxfeepercent` everywhere, which is clearer.

**Comment by ZmnSCPxj** [comment]
Updated as requested, also update manpage.

**Comment by ZmnSCPxj** [comment]
In retrospect, we should probably check that `maxfeepercent` is within the range `0.0 < maxfeepercent <= 100.0`.

**Comment by ZmnSCPxj** [comment]
And also if we use `maxfeepercent` we should probably use `feepercent` and not `fee_percent`.

**Comment by cdecker** [comment]
Renaming `fee_percent` to `feepercent` would be nice, otherwise I'm really happy with this addition

**Comment by ZmnSCPxj** [comment]
Done, ready for review again.

**Comment by cdecker** [comment]
ACK 7f2e594

File: lightningd/pay.c [hunk 2]
```diff
@@ -751,6 +790,25 @@ static void json_pay(struct command *cmd,
...
 	}
 
+	if (maxfeetok
+	    && !json_tok_double(buffer, maxfeetok, &maxfeepercent)) {
+		command_fail(cmd, "'%.*s' is not a valid double",
+			     (int)(maxfeetok->end - maxfeetok->start),
+			     buffer + maxfeetok->start);
+		return;
+	}
+	if (!(0.0 < maxfeepercent)) {
```
Comment (rustyrussell) [inline lightningd/pay.c]: Specifying 0 as maxfeepercent allows you to tell it only to use direct channels, so I think that should be allowed.

And if there's a reason for the test inversion (rounding errors?) please add a comment for those of us who are less wise!

Comment (ZmnSCPxj) [reply lightningd/pay.c]: Not really, I am simply more used to thinking of ranges in the form "min < val < max", and the inversion of the test is simply "not in range".

**Comment by ZmnSCPxj** [comment]
Allow 0.0, also added comment as per requested.

**Comment by cdecker** [comment]
ACK ef95731

**Pull request merged**